File: sdk/typescript/examples/03-governance.ts
/**
 * Example 3: Governance Participation
 * 
 * Demonstrates:
 * - Viewing active proposals
 * - Creating a proposal
 * - Voting on proposals
 * - Checking proposal status
 */

import { AetherClient, AetherKeypair, GovernanceHelper } from '@aether/sdk';

async function main() {
  const client = new AetherClient({ rpcUrl: 'http://localhost:8545' });
  const keypair = await AetherKeypair.fromSeed('my seed phrase');

  const gov = new GovernanceHelper(client, keypair);

  console.log('Fetching active proposals...');
  const proposals = await gov.getActiveProposals();
  
  console.log(`Found ${proposals.length} active proposals\n`);
  
  for (const p of proposals) {
    console.log(`Proposal #${p.id}: ${p.title}`);
    console.log(`- Status: ${p.status}`);
    console.log(`- Votes For: ${p.votesFor}`);
    console.log(`- Votes Against: ${p.votesAgainst}`);
    
    const status = await gov.getProposalStatus(p.id);
    if (status) {
      console.log(`- Has Quorum: ${status.hasQuorum}`);
      console.log(`- Time Remaining: ${status.timeRemaining} slots`);
    }
    console.log();
  }

  console.log('Creating a new proposal...');
  const createTx = await gov.createProposal(
    'Increase Validator Rewards',
    'Proposal to increase validator block rewards from 10 AIC to 15 AIC to improve network security',
    100800
  );
  
  const createHash = await client.sendTransaction(createTx);
  console.log('Proposal created:', createHash);
  await client.waitForTransaction(createHash);

  if (proposals.length > 0) {
    const proposalId = proposals[0].id;
    console.log(`\nVoting on proposal #${proposalId}...`);
    
    const votingPower = await gov.getVotingPower(keypair.address);
    console.log('Your voting power:', votingPower.toString());
    
    const voteTx = await gov.vote(proposalId, true);
    const voteHash = await client.sendTransaction(voteTx);
    console.log('Vote submitted:', voteHash);
    await client.waitForTransaction(voteHash);
    console.log('Vote confirmed');

    const myVote = await gov.getVote(proposalId, keypair.address);
    console.log('\nYour vote:');
    console.log('- Support:', myVote?.support);
    console.log('- Power:', myVote?.votingPower.toString());
  }
}

main().catch(console.error);

File: sdk/python/examples/03_governance.py
"""
Example 3: Governance Participation

Demonstrates:
- Viewing active proposals
- Creating a proposal
- Voting on proposals
- Checking proposal status
"""

import asyncio
from aether import AetherClient, Keypair, GovernanceHelper


async def main():
    async with AetherClient(rpc_url="http://localhost:8545") as client:
        keypair = Keypair.from_seed("my seed phrase")

        gov = GovernanceHelper(client, keypair)

        print("Fetching active proposals...")
        proposals = await gov.get_active_proposals()

        print(f"Found {len(proposals)} active proposals\n")

        for p in proposals:
            print(f"Proposal #{p.id}: {p.title}")
            print(f"- Status: {p.status}")
            print(f"- Votes For: {p.votes_for}")
            print(f"- Votes Against: {p.votes_against}")

            status = await gov.get_proposal_status(p.id)
            if status:
                print(f"- Has Quorum: {status['has_quorum']}")
                print(f"- Time Remaining: {status['time_remaining']} slots")
            print()

        print("Creating a new proposal...")
        create_tx = await gov.create_proposal(
            "Increase Validator Rewards",
            "Proposal to increase validator block rewards from 10 AIC to 15 AIC",
            100800
        )

        create_hash = await client.send_transaction(create_tx)
        print(f"Proposal created: {create_hash}")
        await client.wait_for_transaction(create_hash)

        if proposals:
            proposal_id = proposals[0].id
            print(f"\nVoting on proposal #{proposal_id}...")

            voting_power = await gov.get_voting_power(keypair.address)
            print(f"Your voting power: {voting_power}")

            vote_tx = await gov.vote(proposal_id, True)
            vote_hash = await client.send_transaction(vote_tx)
            print(f"Vote submitted: {vote_hash}")
            await client.wait_for_transaction(vote_hash)
            print("Vote confirmed")

            my_vote = await gov.get_vote(proposal_id, keypair.address)
            if my_vote:
                print("\nYour vote:")
                print(f"- Support: {my_vote.support}")
                print(f"- Power: {my_vote.voting_power}")


if __name__ == "__main__":
    asyncio.run(main())

